include/nlohmann/json.hpp

```diff
@@ -3693,11 +3706,66 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
         // operator[] only works for objects
         if (JSON_HEDLEY_LIKELY(is_object()))
         {
-            return set_parent(m_value.object->operator[](key));
+            return set_parent(m_value.object->operator[](std::forward<KeyT>(key)));
+        }
+
+        JSON_THROW(type_error::create(305, "cannot use operator[] with a string argument with " + std::string(type_name()), *this));
+    }
+
+#if defined(JSON_HAS_CPP_17)
+    /*!
+    @brief access specified object element
+
+    Returns a reference to the element at with specified key @a key.
+
+    @note If @a key is not found in the object, then it is silently added to
+    the object and filled with a `null` value to make `key` a valid reference.
+    In case the value was `null` before, it is converted to an object.
+
+    @param[in] key  key of the element to access
+
+    @return reference to the element at key @a key
+
+    @throw type_error.305 if the JSON value is not an object or null; in that
+    cases, using the [] operator with a key makes no sense.
+
+    @complexity Logarithmic in the size of the container.
+
+    @liveexample{The example below shows how object elements can be read and
+    written using the `[]` operator.,operatorarray__key_type}
+
+    @sa see @ref at(const typename object_t::key_type&) for access by reference
+    with range checking
+    @sa see @ref value() for access by value with a default value
+
+    @since version 3.10.0
+    */
+    reference operator[](const std::string_view& key)
+    {
+        // implicitly convert null value to an empty object
+        if (is_null())
+        {
+            m_type = value_t::object;
+            m_value.object = create<object_t>();
+            assert_invariant();
+        }
+
+        // operator[] only works for objects
+        if (JSON_HEDLEY_LIKELY(is_object()))
+        {
+            auto it = m_value.object->find(key);
+            if (it != m_value.object->end())
+            {
+                return it->second;
+            }
+
+            auto result = m_value.object->emplace(key, nullptr);
+            return set_parent(result.first->second);
```

Comment:
you probably want structured bindings for this.
```cpp
auto [it, success] = m_value.object->emplace(key, nullptr);
assert(success);
return set_parent(it->second);
```

Reply:
I would like to avoid using C++17 style syntax as it looks weird in a 99.9% C++11 code base.

Reply:
If `std::map::find` failed to find the key in the map, how could the assertion ever take effect?

Reply:
> If `std::map::find` failed to find the key in the map, how could the assertion ever take effect?
> I guess that would be the point of assertions, but it is an impossible scenario, thus an unreachable assertion branch, wouldn't it be?

Force of habit :-)